### PR TITLE
Update PyTorch ecosystem

### PR DESCRIPTION
**Pull request by adamjstewart**
* https://github.com/pytorch/pytorch/releases/tag/v2.1.0
* https://github.com/pytorch/audio/releases/tag/v2.1.0
* https://github.com/pytorch/text/releases/tag/v0.16.0
* https://github.com/pytorch/vision/releases/tag/v0.16.0

### CODE REVIEW

**Comment by adamjstewart** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/py-torchtext/package.py [hunk 1]
```diff
@@ -8,67 +8,37 @@
...
+    version("0.5.0", tag="0.5.0", commit="0169cde2f1d446ae886ef0be07e9a673585ed256")
 
     # https://github.com/pytorch/text#installation
     depends_on("python@3.8:3.11", when="@2:", type=("build", "link", "run"))
-    depends_on("python@3.7:3.10", when="@0.13:1", type=("build", "link", "run"))
-    depends_on("python@3.6:3.9", when="@0.8.1:0.12", type=("build", "link", "run"))
-    depends_on("python@3.6:3.8", when="@0.7:0.8.0", type=("build", "link", "run"))
-    depends_on("python@3.5:3.8", when="@0.6", type=("build", "link", "run"))
-    depends_on("python@2.7,3.5:3.8", when="@:0.5", type=("build", "link", "run"))
+    depends_on("python@:3.10", when="@0.13:1", type=("build", "link", "run"))
```
Comment (manuelakuhn) [inline var/spack/repos/builtin/packages/py-torchtext/package.py]:
```
- depends_on("python@3.8:3.11", when="@2:", type=("build", "link", "run"))
- depends_on("python@:3.10", when="@0.13:1", type=("build", "link", "run"))
+ depends_on("python@3.8:3.11", when="@0.15:", type=("build", "link", "run"))
+ depends_on("python@:3.10", when="@0.13:0.14", type=("build", "link", "run"))
```

**Review by manuelakuhn — COMMENTED** [review]
`py-pybind11`

```
+ conflicts("%intel@:17", when="@2.6:")
```
according to [README#L126](https://github.com/pybind/pybind11/blob/v2.11.1/README.rst?plain=1#L126)

`py-torch`

- in `py-project.toml`: `ninja` and `cmake` -> do we need the `py-cmake` and `py-ninja` packages here?
- `py-typing-extensions@3.6.2.1:` -> where does the version restriction come from?
- Why are the dependencies from `pyproject.toml` all `("build", "run")` and not `"build"` dependencies?

I don't understand this package and where the rest of the dependencies come from, so I just believe you :-)

When building with `--test=root``, I get the errors:
```
ModuleNotFoundError: No module named 'packaging'
ModuleNotFoundError: No module named 'expecttest'
```
It did not follow it through with all dependencies but after that I got 
```
Missing pip dependency: pytest-rerunfailures, please run `pip install -r .ci/docker/requirements-ci.txt`
```
The list in [requirements.ci.txt](https://github.com/pytorch/pytorch/blob/v2.1.0/.ci/docker/requirements-ci.txt) is long, so I stopped testing.

Without `--test=root` the package installs fine.

`py-torchtext`

When I try to install it, I get
```
» spack install py-torchtext@0.16                              
==> Error: concretization failed for the following reasons:

   1. Cannot select a single "version" for package "py-torch"
   2. Cannot satisfy 'py-torch@main'
   3. Cannot satisfy 'py-torch@2.1.0'
```
I guess the reason is because it depends on `py-torchdata@0.15:` which still only has a dependency on `py-torch@2.0.1`

**Comment by adamjstewart** [comment]
@aweits @tgaddair @thomas-bouvier any thoughts on the failing horovod build in CI?

**Comment by adamjstewart** [comment]
@manuelakuhn answers to questions:

> in py-project.toml: ninja and cmake -> do we need the py-cmake and py-ninja packages here?

py-cmake and py-ninja are just wrappers around cmake and ninja that Python packages can use to ensure that those non-Python deps are installed. However, for some weird reason, those wrappers cause issues and I've never gotten them to work correctly when building packages. py-blosc2 or py-pybind11 are probably smaller packages if you want to try to reproduce this.

> py-typing-extensions@3.6.2.1: -> where does the version restriction come from?

This comes from https://github.com/onnx/onnx/blob/v1.14.1/requirements.txt. PyTorch vendors ONNX, so any constraints on ONNX also affect PyTorch. Wish we could unvendor it but I got stuck last time I tried: https://github.com/pytorch/pytorch/issues/60331

> Why are the dependencies from pyproject.toml all ("build", "run") and not "build" dependencies?

Because Meta doesn't know how to properly declare dependencies of Python libraries. It's possible everything in `requirements.txt` is required, at least for some part of the library. I wish there was a more clear way to see all required and optional deps but I haven't yet found anything in the docs.

**Review by manuelakuhn — APPROVED** [review]
Thanks for the explanation!

**Comment by aweits** [comment]
> @aweits @tgaddair @thomas-bouvier any thoughts on the failing horovod build in CI?

horovod/torch/CMakeLists.txt:

```
if(NOT Pytorch_VERSION VERSION_LESS "1.5.0")
    set(CMAKE_CXX_STANDARD 14)
endif()
```
changing this to 17 allows it to build...

**Comment by adamjstewart** [comment]
Thanks for figuring this out @aweits. Maybe @tgaddair @thomas-bouvier can fix this upstream and we can backport a patch. Or just wait for the next release.

Anyway, wanted to merge now so people can start using PyTorch 2.1.

**Comment by thomas-bouvier** [comment]
I will handle that!

**Pull request merged**